src/ray/raylet/scheduling/fair_scheduling_queue.cc

```diff
@@ -0,0 +1,113 @@
+#include "ray/raylet/scheduling/fair_scheduling_queue.h"
+
+namespace ray {
+namespace raylet {
+
+using namespace internal;
+
+FairSchedulingQueue::FairSchedulingQueue()
+    : comparator_(active_tasks_), work_queue_(comparator_) {}
+
+void FairSchedulingQueue::Push(const Work &work) {
+  const auto scheduling_class =
+      std::get<0>(work).GetTaskSpecification().GetSchedulingClass();
+  work_queue_[scheduling_class].push_back(work);
+}
+
+void FairSchedulingQueue::PushAll(const SchedulingClass scheduling_class,
+                                  std::deque<Work> works) {
+  // TODO (Alex): We may want to do an O(1) move here if this gets too
+  // inefficient.
+  for (const auto &work : works) {
+    Push(work);
+  }
+}
+
+void FairSchedulingQueue::MarkRunning(const Task &task) {
+  // NOTE: It's unsafe to change `active_works_[scheduling_class]` while
+  // `scheduling_class` is in `work_queue_` since `works_queue_`'s comparator
+  // relies on `active_works_` and thus may become inconsistent.
+  const auto scheduling_class = task.GetTaskSpecification().GetSchedulingClass();
+
+  auto it = work_queue_.find(scheduling_class);
+  if (it != work_queue_.end()) {
+    auto queue = std::move(it->second);
+    work_queue_.erase(scheduling_class);
+    active_tasks_[scheduling_class]++;
+    work_queue_.emplace(scheduling_class, queue);
+  } else {
+    active_tasks_[scheduling_class]++;
```

Comment:
Move this part and the one in the following function into a separate function:

`AdjustPriority(key, int diff)`